chart/compass/charts/director/files/certificate_rotation.sh

```diff
@@ -0,0 +1,119 @@
+RED='\033[0;31m'
+GREEN='\033[0;32m'
+YELLOW='\033[0;33m'
+NC='\033[0m' # No Color
+
+set -e
+
+IS_CERT_ISSUER_LOCALITY_VALID=false
+
+trap "echo '# KILLING PILOT-AGENT #'; curl -X POST http://127.0.0.1:15020/quitquitquit; sleep 3;" RETURN EXIT INT TERM
+
+function getClientCert() {
+  echo -e "${YELLOW}Getting the certificate chain... ${NC}"
+  CERT_CHAIN_PKCS7_RESP=$(curl -s -m 30 -X POST \
+    "$CERT_SVC_CSR_ENDPOINT$CERT_SVC_API_PATH" \
+    -H "Authorization: Bearer $TOKEN" \
+    -H 'Content-Type: application/json' \
+    -H 'Accept: application/json' \
+    -d "{
+          \"certificate-signing-request\": {
+              \"value\": $JSON_CSR,
+              \"type\": \"pkcs10-pem\",
+              \"validity\": {
+                \"value\": $CERT_VALIDITY,
+                \"type\": \"DAYS\"
+            }
+          }
+        }")
+
+  echo -e "${YELLOW}JSON-decoding client certificate chain... ${NC}"
+  CERT_CHAIN_PKCS7=$(jq -r '.["certificate-response"]["value"]' <<< "$CERT_CHAIN_PKCS7_RESP")
+
+  if [ "$CERT_CHAIN_PKCS7" == "null" ]
+  then
+    echo -e "${RED}Could not get certificate response. Reason: ${NC}"
+    echo "$CERT_CHAIN_PKCS7_RESP"
+    exit 1
+  fi
+
+  echo -e "${YELLOW}Extracting client certificate... ${NC}"
+  openssl pkcs7 -print_certs -out /tmp/client-certificate_pkcs7.pem <<< "$CERT_CHAIN_PKCS7"
+  openssl x509 -in /tmp/client-certificate_pkcs7.pem -out /tmp/client-certificate.pem
+}
+
+function confirmValidIssuerLocalityOrRetry() {
+  for (( i = 0; i < "$CLIENT_CERT_RETRY_ATTEMPTS"; i++ )); do
+    echo -e "${YELLOW}Checking issuer locality... ${NC}"
+    ISSUER_LOCALITY=$(openssl x509 -in /tmp/client-certificate.pem -noout -text | grep "Issuer:" | awk '{print $7}' | cut -d '=' -f2 | sed 's/,$//g')
+
+    ARRAY_OF_LOCALITIES=($(echo "$EXPECTED_ISSUER_LOCALITY" | tr ',' '\n'))
+
+    for LOCALITY in ${ARRAY_OF_LOCALITIES[@]}
+    do
+      if [[ "$ISSUER_LOCALITY" == "$LOCALITY" ]]; then
+        echo -e "${GREEN}Issuer locality of the client certificate is valid. Proceeding with the next steps... ${NC}"
+        IS_CERT_ISSUER_LOCALITY_VALID=true
+        break 2
+      fi
+    done
+
+    echo -e "${RED}The issuer locality of the client certificate didn't match any of the expected ones. We expect one of \"$EXPECTED_ISSUER_LOCALITY\" but have: \"$ISSUER_LOCALITY\" ${NC}"
+    echo -e "${YELLOW}[Retry $(($i+1))] Getting new client certificate... ${NC}"
+    getClientCert # This will override the client certificate file content with the newly issued certificate
+    sleep 0.2 # Sleep for 200ms before next retry
+  done
+
+  if [[ $IS_CERT_ISSUER_LOCALITY_VALID == false ]]; then
+    echo -e "${RED}Couldn't get the client certificate with valid issuer locality after $CLIENT_CERT_RETRY_ATTEMPTS attempts. Exiting... ${NC}"
+    exit 1
+  fi
+}
+
+echo -e "${YELLOW}Issuing token... ${NC}"
+
+if [[ "$EXT_NOT_EXTERNAL" == true ]]; then
```

Comment:
What's the meaning behind this `EXT_NOT_EXTERNAL` variable? As far as I can see this var is used to determine if we should encode the certificate or not when saving it to a file? If that's the situation could we use another name for it?
Also, did you try to "unify" both approaches? Either to encode the cert/key for both jobs or both jobs to use non-encoded data? Seems the cert/key files are used only when getting tokens and from the scripts it seems both approaches are working? Is there a specific reason for this encoding?

Reply:
I tried unifying and it did not work. I could not understand why one is encoded and the other not, I assumed it is an implementation detail

Reply:
strange, nevertheless, could we still use a different name for the variable? e.g. `should_encode_cert`? or something similar?